app/src/main/kotlin/com/github/gotify/CoilHandler.kt

```diff
@@ -0,0 +1,62 @@
+package com.github.gotify
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.graphics.drawable.BitmapDrawable
+import coil.ImageLoader
+import coil.annotation.ExperimentalCoilApi
+import coil.disk.DiskCache
+import coil.request.ImageRequest
+import com.github.gotify.api.CertUtils
+import com.github.gotify.client.model.Application
+import okhttp3.OkHttpClient
+import org.tinylog.kotlin.Logger
+import java.io.IOException
+
+internal class CoilHandler(private val context: Context, private val settings: Settings) {
+    private val imageLoader = makeImageLoader()
+
+    private fun makeImageLoader(): ImageLoader {
+        val builder = OkHttpClient.Builder()
```

Comment:
In my testing I can't get the cache headers to get recognized. I'm using caddy with this config:
```caddy
{
    http_port 8000
}

http://192.168.178.2:8000 {
    header Cache-Control "no-cache, no-store, must-revalidate"
    header Vary *
    log
    file_server browse {
        root /home/jm/src/playground/images/
    }
}
```
The served image isn't refetched when I push a new message containing the same image. It seems like the cache-header affects whether the response is put into the disk cache, but it's added to the in emory cache regardless. If the memory cache is disabled via
```
        return ImageLoader.Builder(context)
            .memoryCachePolicy(CachePolicy.READ_ONLY)
```
It seems to work correctly. With cache-control like this:
```
    header Cache-Control "max-age=30, must-revalidate"
```
the image is only refetched after 30s when a new message is pushed.

I'm not sure what to do here. The current solution works similar as with Picasso, but disabling the in memory cache is inefficient. Do you know anything about this or have ideas what to do?

Reply:
I have no simple way to test the caching by myself.  
The memory cache should be cleared when relaunching the app, right?

Because then I don't see a problem here.

Reply:
> The memory cache should be cleared when relaunching the app, right?

The memory cache is currently initialized in the viewmodel of the MessageActivity via the CoilHandler. The memory cache will only be cleared/reinitialized when the MessageActivity is re-created. Meaning only a resume shouldn't be enough.

I'm not sure when this happens because gotify shouldn't be killed because of disabled battery optimization.

Reply:
Sorry, what I mean was closing the app and at some point starting it again. That should invalidate the viewmodel.  
At least it seems to work for me.

But let me test that again with caddy by myself.

Reply:
Removing it from the recent apps thingy will invalidate it but switching to another app isn't enough.

Maybe we could also clear the image cache when swiping downwards. In com.github.gotify.CoilHandler#evict the memory cache should be cleared too.

Reply:
I finally had the time to test it on my own with Caddy. Your results are reproducible.

After some testing I can say:
- just evicting the memory cache is not enough, as the already shown images will not be rerendered
- the diff util does not trigger a refresh because the message id and image url do not change ([mentioned code here](https://github.com/gotify/android/blob/286074386d81e00da6833815add6cf0faeaa0e46/app/src/main/kotlin/com/github/gotify/messages/ListMessageAdapter.kt#L187-L199))
- when forcing the recyclerview to refresh on swiping downwards it works, but it's not really elegant

My preferred solution would be to let the diff util detect new images by itself and only update those entries in the list, but unfortunately we would need an attribute inside the Message objects, which would indicate a difference between images.

Do you have any other ideas?

Reply:
I think we can ignore the possibility that the image updates for an already existing message. Gotify messages are immutable, so I guess we can take it as given, that a message content is also immutable and doesn't have to be refetched.

If this only affects the gotify refresh / swipe down feature, then I'd say the impact is negligible, and we can rerender all messages. So basically invalidate the recycleview state.

Reply:
You're right, this only affects the refresh feature.  
I will try to find a better solution. If I don't, I will just invalidate the recyclerview as you said.